File: leetcode/133.clone-graph.py
# @leet start
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""


class Solution:
    def cloneGraph(self, node: Node | None) -> Node | None:
        if not node:
            return None

        val_to_clone = {node.val: Node(node.val)}
        stack = [node]

        while stack:
            curr = stack.pop()
            clone = val_to_clone[curr.val]
            clone_neighbors = []

            for adj_node in curr.neighbors:
                if adj_node.val not in val_to_clone:
                    val_to_clone[adj_node.val] = Node(adj_node.val)
                    stack.append(adj_node)

                clone_neighbors.append(val_to_clone[adj_node.val])

            clone.neighbors = clone_neighbors

        return val_to_clone[node.val]


# @leet end

File: leetcode/133.clone-graph.cpp
// @leet start
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
#include <queue>
#include <unordered_map>

class Solution {
 public:
  Node* cloneGraph(Node* node) {
    if (!node) {
      return nullptr;
    }

    std::unordered_map<int, Node*> idxToCloneMap;
    idxToCloneMap[node->val] = new Node(node->val);

    std::queue<Node*> q;
    q.push(node);

    while (!q.empty()) {
      const auto* curr = q.front();
      auto* currClone = idxToCloneMap[curr->val];

      q.pop();

      for (auto* adj : curr->neighbors) {
        if (!idxToCloneMap.contains(adj->val)) {
          idxToCloneMap[adj->val] = new Node(adj->val);
          q.push(adj);
        }

        currClone->neighbors.push_back(idxToCloneMap[adj->val]);
      }
    }

    return idxToCloneMap[node->val];
  }
};
// @leet end
